File: apps/android/app/src/main/java/com/healthmd/domain/registry/HealthMdCoreRegistryAdapter.kt
package com.healthmd.domain.registry

import com.healthmd.core.CoreMetricRegistryProfile
import com.healthmd.core.CoreMetricRegistrySnapshot
import com.healthmd.core.HealthMdCoreService
import com.healthmd.domain.model.HealthDataFields
import com.healthmd.domain.model.HealthMetricCategory
import com.healthmd.domain.model.HealthMetricDefinition
import com.healthmd.domain.model.HealthMetrics
import com.healthmd.domain.model.HEALTHMD_CORE_REGISTRY_SHA256

/**
 * Native presentation/capability adapter for the shared Rust registry.
 *
 * Health Connect record classes, feature detection, permissions, persistence, and localized
 * resources remain Kotlin-owned. One coarse immutable snapshot is loaded and cached by callers.
 */
object HealthMdCoreRegistryAdapter {
    fun snapshot(
        profile: CoreMetricRegistryProfile,
        service: HealthMdCoreService = HealthMdCoreService(),
    ): CoreMetricRegistrySnapshot = service.getMetricRegistry(profile)

    fun definitions(snapshot: CoreMetricRegistrySnapshot): List<HealthMetricDefinition> {
        when (snapshot.profileId) {
            "android_frozen_v4" -> {
                require(snapshot.publicProfileId == "android-frozen-v4")
                require(snapshot.publicSchemaVersion == 4u)
            }
            "android_analytical_v5" -> {
                require(snapshot.publicProfileId == "android-analytical-v5")
                require(snapshot.publicSchemaVersion == 5u)
            }
            else -> error("Unsupported shared-core registry profile")
        }
        require(snapshot.publicSchema == "healthmd.health_data")
        require(snapshot.registryVersion == 1u)
        require(snapshot.profileRevision == 1u)
        require(snapshot.registrySha256 == HEALTHMD_CORE_REGISTRY_SHA256)
        require(snapshot.metrics.size == 106)
        return snapshot.metrics.mapIndexed { index, metric ->
            require(metric.ordinal == index.toUInt())
            HealthMetricDefinition(
                id = metric.selectionId,
                category = category(metric.categoryId),
                unit = metric.unit,
            )
        }
    }

    /** Health-free JSON-pointer-style differences for shadow rollout diagnostics. */
    fun shadowDifferences(
        snapshot: CoreMetricRegistrySnapshot,
        nativeDefinitions: List<HealthMetricDefinition> = HealthMetrics.allMetrics,
    ): List<String> = buildList {
        val rustDefinitions = runCatching { definitions(snapshot) }.getOrElse {
            add("/registry/invalid_metadata")
            return@buildList
        }
        if (rustDefinitions.size != nativeDefinitions.size) add("/metrics/count")
        rustDefinitions.zip(nativeDefinitions).forEachIndexed { index, (rust, native) ->
            val prefix = "/metrics/$index"
            if (rust.id != native.id) add("$prefix/selection_id")
            if (rust.category != native.category) add("$prefix/category_id")
            if (rust.unit != native.unit) add("$prefix/unit")
        }
        val nativeUnavailable = HealthMetrics.unavailableMetrics
        if (snapshot.unavailableMetrics.size != nativeUnavailable.size) {
            add("/unavailable_metrics/count")
        }
        snapshot.unavailableMetrics.zip(nativeUnavailable).forEachIndexed { index, (rust, native) ->
            val prefix = "/unavailable_metrics/$index"
            if (rust.selectionId != native.id) add("$prefix/selection_id")
            if (category(rust.categoryId) != native.category) add("$prefix/category_id")
            if (rust.labelKey != native.id) add("$prefix/label_key")
            if (rust.reasonKey != "metric_unavailable_${native.id}") add("$prefix/reason_key")
        }
        if (snapshot.outputs.map { it.key } != HealthDataFields.allKeys) add("/outputs/flat")
        if (snapshot.outputs.map { it.ordinal } != snapshot.outputs.indices.map { it.toUInt() }) {
            add("/outputs/order")
        }
    }

    private fun category(value: String): HealthMetricCategory = when (value) {
        "Sleep" -> HealthMetricCategory.SLEEP
        "Activity" -> HealthMetricCategory.ACTIVITY
        "Heart" -> HealthMetricCategory.HEART
        "Respiratory" -> HealthMetricCategory.RESPIRATORY
        "Vitals" -> HealthMetricCategory.VITALS
        "Body" -> HealthMetricCategory.BODY
        "Nutrition" -> HealthMetricCategory.NUTRITION
        "Mobility" -> HealthMetricCategory.MOBILITY
        "Cycling" -> HealthMetricCategory.CYCLING
        "Hearing" -> HealthMetricCategory.HEARING
        "Mindfulness" -> HealthMetricCategory.MINDFULNESS
        "Reproductive" -> HealthMetricCategory.REPRODUCTIVE
        "Symptoms" -> HealthMetricCategory.SYMPTOMS
        "Medications" -> HealthMetricCategory.MEDICATIONS
        "Other" -> HealthMetricCategory.OTHER
        "Workouts" -> HealthMetricCategory.WORKOUTS
        else -> error("Invalid shared-core category metadata")
    }
}

File: apps/apple/HealthMd/Shared/Core/HealthMdCoreRegistryAdapter.swift
import Foundation
import HealthMdCoreRust

/// Native presentation/SDK adapter for the shared Rust registry.
///
/// HealthKit type resolution, `#available` evaluation, permissions, persistence, and localized
/// strings remain in Swift. The adapter consumes one coarse immutable profile snapshot.
enum HealthMdCoreRegistryAdapter {
    enum AdapterError: Error, Equatable {
        case invalidMetadata
    }

    static func appleSnapshot(
        service: HealthMdCoreService = HealthMdCoreService()
    ) throws -> CoreMetricRegistrySnapshot {
        try service.metricRegistry(profile: .appleHealthDataV8)
    }

    static func definitions(
        from snapshot: CoreMetricRegistrySnapshot
    ) throws -> [HealthMetricDefinition] {
        guard snapshot.profileId == "apple_health_data_v8",
              snapshot.publicProfileId == "apple-v8",
              snapshot.publicSchema == HealthMdExportSchema.identifier,
              snapshot.publicSchemaVersion == UInt32(HealthMdExportSchema.version),
              snapshot.registryVersion == 1,
              snapshot.profileRevision == 1,
              snapshot.registrySha256 == HealthMetrics.registrySHA256,
              snapshot.metrics.count == 230 else {
            throw AdapterError.invalidMetadata
        }

        return try snapshot.metrics.enumerated().map { index, metric in
            guard metric.ordinal == UInt32(index),
                  let category = HealthMetricCategory(rawValue: metric.categoryId),
                  let availability = HealthMetricAvailability(rawValue: metric.availabilityKey),
                  let metricType = metricType(metric.kind),
                  let aggregation = aggregation(metric.sourceAggregation) else {
                throw AdapterError.invalidMetadata
            }
            return HealthMetricDefinition(
                id: metric.selectionId,
                name: metric.referenceName,
                category: category,
                unit: metric.unit,
                healthKitIdentifier: metric.sourceSelector == "None" ? nil : metric.sourceSelector,
                metricType: metricType,
                aggregation: aggregation,
                isArchiveOnly: metric.archiveOnly,
                isEnabledByDefault: metric.defaultEnabled,
                availability: availability
            )
        }
    }

    /// Health-free JSON-pointer-style differences for shadow rollout diagnostics.
    static func shadowDifferences(
        snapshot: CoreMetricRegistrySnapshot,
        nativeDefinitions: [HealthMetricDefinition] = HealthMetrics.all
    ) -> [String] {
        var differences: [String] = []
        guard let rustDefinitions = try? definitions(from: snapshot) else {
            return ["/registry/invalid_metadata"]
        }
        if rustDefinitions.count != nativeDefinitions.count {
            differences.append("/metrics/count")
        }
        for index in 0..<min(rustDefinitions.count, nativeDefinitions.count) {
            let rust = rustDefinitions[index]
            let native = nativeDefinitions[index]
            let prefix = "/metrics/\(index)"
            if rust.id != native.id { differences.append("\(prefix)/selection_id") }
            if rust.name != native.name { differences.append("\(prefix)/reference_name") }
            if rust.category != native.category { differences.append("\(prefix)/category_id") }
            if rust.unit != native.unit { differences.append("\(prefix)/unit") }
            if rust.healthKitIdentifier != native.healthKitIdentifier { differences.append("\(prefix)/source_selector") }
            if metricTypeKey(rust.metricType) != metricTypeKey(native.metricType) {
                differences.append("\(prefix)/kind")
            }
            if aggregationKey(rust.aggregation) != aggregationKey(native.aggregation) {
                differences.append("\(prefix)/source_aggregation")
            }
            if rust.isArchiveOnly != native.isArchiveOnly { differences.append("\(prefix)/archive_only") }
            if rust.isEnabledByDefault != native.isEnabledByDefault { differences.append("\(prefix)/default_enabled") }
            if rust.availability != native.availability { differences.append("\(prefix)/availability_key") }
        }

        let rustArchive = Set(snapshot.metrics.filter(\.archiveOnly).map(\.selectionId))
        if rustArchive != HealthMetricExportMapping.reviewedArchiveOnlyMetricIDs {
            differences.append("/outputs/archive_only")
        }
        let rustMapping = Dictionary(
            grouping: snapshot.outputs,
            by: { $0.selectionIds.first ?? "" }
        ).mapValues { outputs in
            outputs.sorted { $0.ordinal < $1.ordinal }.map(\.key)
        }
        if rustMapping != HealthMetricExportMapping.metricIdToFrontmatterKeys {
            differences.append("/outputs/flat")
        }
        return differences
    }

    private static func metricTypeKey(_ value: HealthMetricDefinition.MetricType) -> String {
        switch value {
        case .quantity: "quantity"
        case .category: "category"
        case .workout: "workout"
        }
    }

    private static func aggregationKey(_ value: HealthMetricDefinition.AggregationType) -> String {
        switch value {
        case .cumulative: "cumulative"
        case .discreteAvg: "discreteAvg"
        case .discreteMin: "discreteMin"
        case .discreteMax: "discreteMax"
        case .mostRecent: "mostRecent"
        case .duration: "duration"
        case .count: "count"
        }
    }

    private static func metricType(
        _ value: String
    ) -> HealthMetricDefinition.MetricType? {
        switch value {
        case "quantity": .quantity
        case "category": .category
        case "workout": .workout
        default: nil
        }
    }

    private static func aggregation(
        _ value: String
    ) -> HealthMetricDefinition.AggregationType? {
        switch value {
        case "cumulative": .cumulative
        case "discreteAvg": .discreteAvg
        case "discreteMin": .discreteMin
        case "discreteMax": .discreteMax
        case "mostRecent": .mostRecent
        case "duration": .duration
        case "count": .count
        default: nil
        }
    }
}
